Unmount recursively to unmount volumes

rhvgoyal opened the pull request:
Fixes #17823 

Bind mounted volumes are bind mounted recursively. That means any mounts
under top mount also becomes visible under destination mount point. And
that also means that simple umount(dest) will fail as it has other mount
points underneath it. Instead use utitlity "umount -R dest" for recursively
unmounting all the mounts under dest.

Signed-off-by: Vivek Goyal vgoyal@redhat.com

rhvgoyal commented:
@cpuguy83 Does this look like a reasonable solution?

rhvgoyal commented:
@bboreham Can you please test this patch and see if it solves your issue.

rhvgoyal commented:
fixed golint issue and pushed again.

LK4D4 commented on pkg/system/syscall_unix.go:
```diff
@@ -2,10 +2,23 @@
+// UnmountRecursive is a platform-specific helper function to call
+// the umount -R command provided by util-linux package.
+func UnmountRecursive(dest string) error {
+	if out, err := exec.Command("umount", "-R", dest).CombinedOutput(); err != nil {
```
I think we used `MNT_DETACH` before for such task...

rhvgoyal committed:
Use lazy unmount to unmount volumes

Volumes can have more mount points beneath them and unmount will fail. This
is the case when a bind mounted directory on host already had a mount point
underneath it. So use lazy unmount instead.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
@crosbymichael @LK4D4 @cpuguy83 Pushed new version doing lazy unmount. This this look better.

crosbymichael commented:
LGTM

LK4D4 commented:
LGTM

anusha-ragunathan commented:
LGTM

cpuguy83 commented:
@rhvgoyal Thanks for looking into this!

rade commented:
any chance to get this into 1.9.2?

LK4D4 commented:
@rade sure, if there will be 1.9.2, why not

thaJeztah commented:
@rade note that it's not been decided yet if we will actually be releasing a 1.9.2 release, we use the milestone to collect issues that may make sense in a 1.9.2 release, but it will only be released if there are serious issues that warrant a release
